TransientObjectException when scheduling a task previously saved

A TransientObjectException exception is thrown when scheduling a task that was previously saved.

To reproduce:
Create new project using a template with two tasks, one scheduled and one not scheduled.
Allocate a resource to the scheduled task.
Save the project.
Edit the project
Add hours to the unscheduled task
Click the schedule button
Switch to project scheduling tab
Allocate a resource to the new scheduled task.
Save the project --> throws exception

ERROR [21-January 12:45:23] [Thread-27] org.hibernate.internal.SessionImpl  - HHH000346: Error during managed flush [object references an unsaved transient instance - save the transient instance before flushing: org.libreplan.business.planner.entities.TaskElement]
ERROR [21-January 12:45:23] [qtp1126888535-108] org.libreplan.web.error.PageForErrorOnEvent  - object references an unsaved transient instance - save the transient instance before flushing: org.libreplan.business.planner.entities.TaskElement; nested exception is org.hibernate.TransientObjectException: object references an unsaved transient instance - save the transient instance before flushing: org.libreplan.business.planner.entities.TaskElement [Status Code: 500]
org.springframework.dao.InvalidDataAccessApiUsageException: object references an unsaved transient instance - save the transient instance before flushing: org.libreplan.business.planner.entities.TaskElement; nested exception is org.hibernate.TransientObjectException: object references an unsaved transient instance - save the transient instance before flushing: org.libreplan.business.planner.entities.TaskElement
	at org.springframework.orm.hibernate5.SessionFactoryUtils.convertHibernateAccessException(SessionFactoryUtils.java:191)
	at org.springframework.orm.hibernate5.HibernateTransactionManager.convertHibernateAccessException(HibernateTransactionManager.java:741)
	at org.springframework.orm.hibernate5.HibernateTransactionManager.doCommit(HibernateTransactionManager.java:589)
	at org.springframework.transaction.support.AbstractPlatformTransactionManager.processCommit(AbstractPlatformTransactionManager.java:761)
	at org.springframework.transaction.support.AbstractPlatformTransactionManager.commit(AbstractPlatformTransactionManager.java:730)
	at org.springframework.transaction.interceptor.TransactionAspectSupport.commitTransactionAfterReturning(TransactionAspectSupport.java:485)
	at org.springframework.transaction.interceptor.TransactionAspectSupport.invokeWithinTransaction(TransactionAspectSupport.java:291)
	at org.springframework.transaction.interceptor.TransactionInterceptor.invoke(TransactionInterceptor.java:96)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:179)
	at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:208)
	at com.sun.proxy.$Proxy197.confirmSave(Unknown Source)
	at org.libreplan.web.planner.taskedition.TaskPropertiesController.setEmailNotificationEntity(TaskPropertiesController.java:806)
	at org.libreplan.web.planner.taskedition.TaskPropertiesController.proceedList(TaskPropertiesController.java:779)
	at org.libreplan.web.planner.taskedition.TaskPropertiesController.emailNotificationAddNew(TaskPropertiesController.java:753)
	at org.libreplan.web.planner.order.SaveCommandBuilder$SaveCommand.doAction(SaveCommandBuilder.java:268)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at org.libreplan.web.common.concurrentdetection.ConcurrentModificationHandling.lambda$handler$462(ConcurrentModificationHandling.java:60)
	at com.sun.proxy.$Proxy258.doAction(Unknown Source)
	at org.zkoss.ganttz.CommandContextualized.doAction(CommandContextualized.java:48)
	at org.zkoss.ganttz.CommandContextualized.lambda$toButton$12(CommandContextualized.java:67)
	at org.zkoss.zk.ui.AbstractComponent.onEvent(AbstractComponent.java:3030)
	at org.zkoss.zk.ui.AbstractComponent.service(AbstractComponent.java:3001)
	at org.zkoss.zk.ui.AbstractComponent.service(AbstractComponent.java:2942)
	at org.zkoss.zk.ui.impl.EventProcessor.process(EventProcessor.java:136)
	at org.zkoss.zk.ui.impl.EventProcessingThreadImpl.process0(EventProcessingThreadImpl.java:536)
	at org.zkoss.zk.ui.impl.EventProcessingThreadImpl.run(EventProcessingThreadImpl.java:463)
Caused by: org.hibernate.TransientObjectException: object references an unsaved transient instance - save the transient instance before flushing: org.libreplan.business.planner.entities.TaskElement
	at org.hibernate.engine.internal.ForeignKeys.getEntityIdentifierIfNotUnsaved(ForeignKeys.java:279)
	at org.hibernate.type.EntityType.getIdentifier(EntityType.java:455)
	at org.hibernate.type.ManyToOneType.isDirty(ManyToOneType.java:281)
	at org.hibernate.type.ManyToOneType.isDirty(ManyToOneType.java:291)
	at org.hibernate.type.TypeHelper.findDirty(TypeHelper.java:296)
	at org.hibernate.persister.entity.AbstractEntityPersister.findDirty(AbstractEntityPersister.java:4137)
	at org.hibernate.event.internal.DefaultFlushEntityEventListener.dirtyCheck(DefaultFlushEntityEventListener.java:528)
	at org.hibernate.event.internal.DefaultFlushEntityEventListener.isUpdateNecessary(DefaultFlushEntityEventListener.java:215)
	at org.hibernate.event.internal.DefaultFlushEntityEventListener.onFlushEntity(DefaultFlushEntityEventListener.java:142)
	at org.hibernate.event.internal.AbstractFlushingEventListener.flushEntities(AbstractFlushingEventListener.java:216)
	at org.hibernate.event.internal.AbstractFlushingEventListener.flushEverythingToExecutions(AbstractFlushingEventListener.java:85)
	at org.hibernate.event.internal.DefaultFlushEventListener.onFlush(DefaultFlushEventListener.java:38)
	at org.hibernate.internal.SessionImpl.flush(SessionImpl.java:1295)
	at org.hibernate.internal.SessionImpl.managedFlush(SessionImpl.java:468)
	at org.hibernate.internal.SessionImpl.flushBeforeTransactionCompletion(SessionImpl.java:3135)
	at org.hibernate.internal.SessionImpl.beforeTransactionCompletion(SessionImpl.java:2352)
	at org.hibernate.engine.jdbc.internal.JdbcCoordinatorImpl.beforeTransactionCompletion(JdbcCoordinatorImpl.java:491)
	at org.hibernate.resource.transaction.backend.jdbc.internal.JdbcResourceLocalTransactionCoordinatorImpl.beforeCompletionCallback(JdbcResourceLocalTransactionCoordinatorImpl.java:147)
	at org.hibernate.resource.transaction.backend.jdbc.internal.JdbcResourceLocalTransactionCoordinatorImpl.access$100(JdbcResourceLocalTransactionCoordinatorImpl.java:38)
	at org.hibernate.resource.transaction.backend.jdbc.internal.JdbcResourceLocalTransactionCoordinatorImpl$TransactionDriverControlImpl.commit(JdbcResourceLocalTransactionCoordinatorImpl.java:231)
	at org.hibernate.engine.transaction.internal.TransactionImpl.commit(TransactionImpl.java:65)
	at org.springframework.orm.hibernate5.HibernateTransactionManager.doCommit(HibernateTransactionManager.java:581)
	... 26 more

--- a/libreplan-business/src/main/java/org/libreplan/business/common/AdHocTransactionService.java
+++ b/libreplan-business/src/main/java/org/libreplan/business/common/AdHocTransactionService.java
@@ -103,6 +103,7 @@ public class AdHocTransactionService implements IAdHocTransactionService {
         };
     }
 
+    @Override
     @Transactional
     public <T> T runOnTransaction(IOnTransaction<T> onTransaction) {
         return onTransaction.execute();

--- a/libreplan-webapp/src/main/java/org/libreplan/web/planner/order/SaveCommandBuilder.java
+++ b/libreplan-webapp/src/main/java/org/libreplan/web/planner/order/SaveCommandBuilder.java
@@ -589,8 +589,18 @@ public class SaveCommandBuilder {
                 removeEmptyConsolidation(taskElement);
                 updateLimitingResourceQueueElementDates(taskElement);
 
-                if (taskElement.getTaskSource() != null && taskElement.getTaskSource().isNewObject())
+                if (taskElement.getTaskSource() != null && taskElement.getTaskSource().isNewObject()) {
                     saveTaskSources(taskElement);
+
+                    // hack to update taskPropertiesController reference to current currentTaskElement
+                    if (SaveCommandBuilder.taskPropertiesController != null) {
+                        TaskElement taskControllerTaskElement = SaveCommandBuilder.taskPropertiesController.getCurrentTaskElement();
+                        if ( taskControllerTaskElement != null && taskControllerTaskElement.getOrderElement() == taskElement.getOrderElement() &&
+                                taskControllerTaskElement.isNewObject()) {
+                            SaveCommandBuilder.taskPropertiesController.setCurrentTaskElement(taskElement);
+                        }
+                    }
+                }
 
                 updateLimitingQueueDependencies(taskElement);
             }

--- a/libreplan-webapp/src/main/java/org/libreplan/web/planner/taskedition/TaskPropertiesController.java
+++ b/libreplan-webapp/src/main/java/org/libreplan/web/planner/taskedition/TaskPropertiesController.java
@@ -826,4 +826,12 @@ public class TaskPropertiesController extends GenericForwardComposer<Component> 
         return this;
     }
 
+    public TaskElement getCurrentTaskElement() {
+        return currentTaskElement;
+    }
+
+    public void setCurrentTaskElement(TaskElement taskElement) {
+        currentTaskElement = taskElement;
+    }
+
 }

